### PR TITLE
Various fixes to wwvegas stuff.

### PR DESCRIPTION
**Pull request by jonwil**
Add missing IMPLEMENT_W3D_POOL macros to Smudge, SmudgeSet and
BoxPrototypeClass.
Add some asserts to CollisionMath.
Fix W3DPrototypeClass to correctly use MemoryPoolObject instead of
W3DMPO.
Add DX8Wrapper::Clamp_Color.
Properly implement W3dUtilityClass.
Add some things to w3d_file.h
Add TextureClass::Get_Level_Description.
Remove MeshGeometryClass constructor hooks and incorrect
IMPLEMENT_W3D_POOL macro.
Add Greatest_Common_Divisor and Least_Commmon_Multiple.
Add NodeCompressedMotionStruct constructor.
Add SimpleFileFactoryClass and defaultFileFactory.
Add DX8Wrapper::Convert_Color_Clamp.
Properly implement all the HashTemplateClass member functions I can
find implementations of.
Properly implement all the HashTemplateIterator member functions.

### CODE REVIEW

File: src/w3d/lib/gcd_lcm.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "gcd_lcm.h"
+
+unsigned int Greatest_Common_Divisor(unsigned int a, unsigned int b)
```
Comment (duncanspumpkin) [inline src/w3d/lib/gcd_lcm.cpp:17]: What are these being used for? I suggest moving them into the header and making them `constexpr` as they can be computed at compile time for any fixed values of a and b.

File: src/w3d/lib/gcd_lcm.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    return d;
+}
+
+unsigned int Least_Common_Multiple(unsigned int a, unsigned int b)
+{
+    if (b) {
+        return b * a / Greatest_Common_Divisor(b, a % b);
+    } else {
+        return 0 / a;
```
Comment (duncanspumpkin) [inline src/w3d/lib/gcd_lcm.cpp]: This looks wrong.

Comment (jonwil) [reply src/w3d/lib/gcd_lcm.cpp]: I took another look at this in a few different binaries and re-did it.

File: src/w3d/lib/gcd_lcm.cpp (hunk 1 above)
Comment (jonwil) [reply src/w3d/lib/gcd_lcm.cpp:17]: Its used by the particle emitter code and the values passed in are not constants. I took another look at this in a few binaries and discovered evidence about how it actually looked so I changed it.

Comment (jonwil) [reply src/w3d/lib/gcd_lcm.cpp:17]: The way these 2 functions are now is how they look in zero hour worldbuilder so they should be good.

**Review by tomsons26 — CHANGES_REQUESTED** [review]
Def too much and too random for a single pr, need splitting up, at minimum to separate commits, but i'd prefer seperate prs

**Pull request closed without merging**